Residual object

YaqiWang opened the pull request:
Close #13609.

This will be useful for the eigen system cleanups.

Note that the second comment went a little further by replacing `computeOffDiagJacobian(unsigned jvar)` with `computeOffDiagJacobian(MooseVariableFEBase & jvar)`. It does not change the interface of `computeQpOffDiagJacobian`. Thus this PR may break applications if applications override `computeOffDiagJacobian`. It should be pretty rare though.

YaqiWang committed:
add residual object base class #13609

YaqiWang committed:
cleaning up compute off-diagonal jacobian after #13609

YaqiWang commented:
I have to give up the further cleanup, simple because too much objects affected... We can create a script to facilitate that in the future.

YaqiWang commented:
This showed good on my mac and on sawtooth. Should be ready for review. @fdkong or @lindsayad.

YaqiWang commented:
One test is failing in debug mode due to an assertion. The two variables are defined on the original mesh and displaced mesh. Somehow the size of the local Jacobian is zero. What puzzles me is that this does not fail before. Any idea?

fdkong commented:
Looks like you can not totally get of that API because some animals are still using it. It is better to just deprecate it.

loganharbour commented:
I'm a fan of this. I will look into it once tests are passing.

fdkong commented on framework/include/base/ResidualObject.h:
```diff
@@ -0,0 +1,118 @@
+  virtual void computeJacobian() = 0;
+
+  /// Computes d-residual / d-jvar... storing the result in Ke.
+  virtual void computeOffDiagJacobian(MooseVariableFEBase & /*jvar*/) {}
```
It might be even better to use `MooseVariableBase` to cover all cases

YaqiWang replied:
That will be too aggressive.

lindsayad replied:
We want the parameter here to be `unsigned int jvar` instead of `MooseVariable.*`. Having this parameter as `MooseVariable.*` can more easily lead to errors like the one you encountered in the displaced integrated boundary condition test. This will require modifying `ComputeFullJacobianThread::computeFaceJacobian` but that's ok

fdkong reviewed (changes requested):
It looks really good for me. If it passed all tests, I would just get it in

fdkong commented:
> One test is failing in debug mode due to an assertion. The two variables are defined on the original mesh and displaced mesh. Somehow the size of the local Jacobian is zero. What puzzles me is that this does not fail before. Any idea?

It look likely a variable is not defined on that interface or element but `computeJacobian` is still involved. We could check if the number of DoFs is nonzero before we call `computeJacobian`

lindsayad commented on framework/src/bcs/IntegratedBC.C:
```diff
@@ -129,28 +129,22 @@ IntegratedBC::computeJacobian()
+  prepareMatrixTag(_assembly, _var.number(), jvar.number());
 
   // This (undisplaced) jvar could potentially yield the wrong phi size if this object is acting
   // on the displaced mesh
-  auto phi_size = _sys.getVariable(_tid, jvar).dofIndices().size();
+  auto phi_size = jvar.dofIndices().size();
```
Did you read this comment before you made this change?

YaqiWang replied:
ah, no. lol.

lindsayad replied:
It is telling you the exact problem that happened

lindsayad replied:
`_sys` is the displaced nonlinear system if `use_displaced_mesh` for that object is `true`, else it is the undisplaced nonlinear system

YaqiWang replied:
There two other places to fix: `NodeFaceConstraint.C ` and `GapHeatTransfer.C`.

lindsayad commented on framework/include/base/ResidualObject.h:
```diff
@@ -0,0 +1,125 @@
+typedef MooseVariableFE<VectorValue<Real>> VectorMooseVariable;
+typedef MooseVariableFE<RealEigenVector> ArrayMooseVariable;
+
+template <>
+InputParameters validParams<KernelBase>();
```
Don't need this here

loganharbour commented:
While there is general agreement on using `ResidualObject`, there is a desired way of satisfying the other issue of accessing tagged objects in `NonlinearEigenSystem`. See https://github.com/idaholab/moose/issues/16125. I will work on this today.

YaqiWang commented:
@lindsayad I think we should keep `void computeOffDiagJacobian(MooseVariableFEBase & /*jvar*/) ` instead of `void computeOffDiagJacobian(unsigned int /*jvar*/)`. The reason is made clearer in `Kernel::computeOffDiagJacobian`, where we need to check the variable type, normal variable or array variable. Basically a full variable reference provides more information like number of components than only the variable number. I do not worry about using the variables in a wrong system because it is handled in the framework. Potentially we can move some common codes in `Kernel::computeOffDiagJacobian` into `ResidualObject::computeOffDiagJacobian`. But I am not going to pursue that in this PR. I did this PR for the eigenvalue executioner cleanup.

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/617940/) on c81b0f9 wanted to post the following:

View the site [here](https://mooseframework.inl.gov/docs/PRs/16117/site)

This comment will be updated on new commits.

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/618304/) on d3bbbb2 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/16117/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/16117/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format ce49bda7824b48564189c01485a400ab609e8763`

lindsayad commented:
>  I do not worry about using the variables in a wrong system because it is handled in the framework.

This is the framework. And clearly it's not handled as evinced by the earlier test error

lindsayad commented:
> The reason is made clearer in `Kernel::computeOffDiagJacobian`, where we need to check the variable type, normal variable or array variable.

You already have this call in the code you referenced:

```c++
    // This (undisplaced) jvar could potentially yield the wrong phi size if this object is acting
    // on the displaced mesh
    auto phi_size = _sys.getVariable(_tid, jvar.number()).dofIndices().size();
```

which we know we need to avoid the error seen with the integrated bcs. So save-off a reference to the variable here in the `Kernel.C` code and then use it within the method instead of using the `jvar` argument

YaqiWang commented:
You mean save off the pointers of all primal variables and use `unsigned int jvar` to retrieve one? It is then possibly better to move the `use_displaced_mesh` parameter down into residual object. I want this PR a minimum change for introducing `ResidualObject`. We can do that later, can we?

YaqiWang commented:
I noticed a few other things in this PR: `IntegratedBC` is using a name `computeJacobianBlock` instead of `computeOffDiagJacobian`; IMO `ODEKernel::computeJacobian()` should be implemented in `ScalarKernel::computeJacobian()`; `NodalConstraint` has different interface for `computeResidual` and `computeJacobian`, etc. Not sure if I should submit a separate issue for these. Just to let you know.

lindsayad commented:
> You mean save off the pointers of all primal variables and use `unsigned int jvar` to retrieve one? It is then possibly better to move the `use_displaced_mesh` parameter down into residual object. I want this PR a minimum change for introducing `ResidualObject`. We can do that later, can we?

No. This is the code right now:
```c++
void
Kernel::computeOffDiagJacobian(MooseVariableFEBase & jvar)
{
  auto jvar_num = jvar.number();
  if (jvar_num == _var.number())
    computeJacobian();
  else
  {
    prepareMatrixTag(_assembly, _var.number(), jvar_num);

    // This (undisplaced) jvar could potentially yield the wrong phi size if this object is acting
    // on the displaced mesh
    auto phi_size = _sys.getVariable(_tid, jvar.number()).dofIndices().size();
```
This is the proposed code:
```
void
Kernel::computeOffDiagJacobian(unsigned int jvar_num)
{
  const auto & jvar = _sys.getVariable(_tid, jvar_num);

  if (jvar_num == _var.number())
    computeJacobian();
  else
  {
    prepareMatrixTag(_assembly, _var.number(), jvar_num);

    // This (undisplaced) jvar could potentially yield the wrong phi size if this object is acting
    // on the displaced mesh
    auto phi_size = jvar.dofIndices().size();
```
The rest of the method would be totally unchanged. You were already talking about deleting an `offDiagJacobian` API. We're just trying to make sure we delete the right one.

YaqiWang commented:
I can argue we pass in the right variable within `ComputeFullJacobianThread`. `SystemBase::getVariable` call is not so light. I still feel the interface using the variable is the right one.

YaqiWang commented:
Note this:
```
Kernel::computeOffDiagJacobian(unsigned int jvar)
{
  mooseDeprecated("The computeOffDiagJacobian method signature has changed. Developers, please "
                  "pass in a MooseVariableFEBase reference instead of the variable number");
```
Someone already thought of this before.

lindsayad commented:
I'm aware so who more qualified to give an opinion than someone who has done it both ways?

lindsayad commented:
> `SystemBase::getVariable` call is not so light. I still feel the interface using the variable is the right one.

Have you seen it show up in profiling?

YaqiWang commented:
Nope and it is probably not important. But that is an indicator how the code should be designed. The loops can have a hold of two sets of variables and pass in the correct one to the residual objects.

lindsayad commented:
So you still want to do a container query

lindsayad commented:
If you provide a clean implementation with `computeOffDiagJacobian(MooseVariable.*)` which does not require any `_sys.getVariable` calls in the method in order to get the correct variable, then I won't shut the door on that API. But it has to be in this PR. I won't accept this as-is with `_sys.getVariable` calls in `computeOffDiagJacobian(MooseVariable.*)` because that simply exposes that we've set up a dangerous way of communicating coupling information which is liable to break when doing coupled undisplaced/displaced simulations.

You've put up a good PR @YaqiWang that I've gone to bat for you for. You've even exposed flaws in a design that I myself created, as you pointed out. So let's use this excellent opportunity that you've created to make things as beautiful as possible.

YaqiWang commented:
This is a good argument. @lindsayad Not a problem to me. I think your argument is good.

YaqiWang commented:
Noticed that: `defineLegacyParams(ResidualObject);` in ResidualObject.C was not deleted.

lindsayad committed:
Use computeOffDiagJacobian(unsigned int) API

This commit also incorporates a few other changes:

- Remove `computeJacobianBlock` from integrated boundary conditions and
  instead use `computeOffDiagJacobian` in order to be consistent across
  all our residual objects. (See `ResidualObject` encourages consistency
  and uniformity)
- `subProblem()` should be `const`. There's absolutely no reason that
  residual objects should be sending out non-const references to their
  `SubProblem`. That violates encapsulation. Instead we add an API to
  `ResidualObject` for preparing shapes, e.g. `void prepareShapes()`.
  Additionally we create a `NeighborResidualObject` that provides an
  additional API `void prepareNeighborShapes()`. `Constraint`,
  `DGKernelBase`, and `InterfaceKernelBase` now all inherit from
  `NeighborResidualObject`. Creation of `NeighborResidualObject` could
  allow for more abstraction and migration of common derived child code
  down into `NeighborResidualObject`. (Note that this line of thinking is
  also true for `ResidualObject` itself. I'm hopeful that eventually all
  save-in code can be moved down into `ResidualObject` (although save-in
  should also be eliminated anyways))

lindsayad committed:
Update modules for computeOffDiagJacobian(unsigned int)

moosebuild commented:
Job [Minimum clang version](https://civet.inl.gov/job/619476/) on 2ffcd47 : invalidated by @loganharbour

Test box failure

lindsayad commented:
@YaqiWang can you explain why you removed `computeQpOffDiagJacobian(unsiged int)` from `ScalarKernel`?

lindsayad commented:
> @YaqiWang can you explain why you removed `computeQpOffDiagJacobian(unsiged int)` from `ScalarKernel`?

I guess this was actually removed from `ODEKernel`. I can see that we only actually do off-diagonal Jacobian computation (looking at `ODEKernel::computeOffDiagJacobian`) for scalar variables, so maybe that was your reasoning...which I think is good reasoning.

lindsayad commented:
Yea to your earlier point, I don't really understand why we have `ScalarKernel` and `ODEKernel` separated, but I guess `GlobalStrain` does inherit from `ScalarKernel` as opposed to `ODEKernel` and it implements much different `computeResidual` and `computeJacobian`

YaqiWang commented:
I think we should clean up that scalar kernel. But we can do it in a separate PR.

YaqiWang commented on framework/src/base/ResidualObject.C:
```diff
@@ -62,3 +54,9 @@ ResidualObject::ResidualObject(const InputParameters & parameters, bool is_nodal
+void
+ResidualObject::prepareShapes(const unsigned int var_num)
+{
+  _subproblem.prepareShapes(var_num, _tid);
```
What the prepare shapes is doing? Are the two kernels having the same thing thus we could be duplicating computation here?

YaqiWang commented on framework/src/kernels/VectorKernel.C:
```diff
@@ -99,24 +99,25 @@ VectorKernel::computeJacobian()
 
 void
-VectorKernel::computeOffDiagJacobian(MooseVariableFEBase & jvar)
+VectorKernel::computeOffDiagJacobian(const unsigned int jvar_num)
```
const unsigned int?

lindsayad replied:
This is to prevent some deviant from coming into the method body and reassigning `jvar_num`.

YaqiWang commented on framework/src/loops/ComputeFullJacobianThread.C:
```diff
@@ -60,8 +60,8 @@ ComputeFullJacobianThread::computeJacobian()
         {
-          kernel->subProblem().prepareShapes(jvar, _tid);
-          kernel->computeOffDiagJacobian(jvariable);
+          kernel->prepareShapes(jvar);
```
Oh, I see here. We are not introducing anything new then.

lindsayad replied:
Right, this is just introducing more encapsulation. I feel like if we have a non-const `subProblem()` getter in our residual objects, we are doing something wrong

YaqiWang commented on framework/include/constraints/Constraint.h:
```diff
@@ -22,7 +22,7 @@ InputParameters validParams<Constraint>();
  * Base class for all Constraint types
  */
-class Constraint : public ResidualObject, protected GeometricSearchInterface
+class Constraint : public NeighborResidualObject, protected GeometricSearchInterface
```
missing NeighborResidualObject source in this commit?

lindsayad replied:
Yea it's in a fixup commit (trying to avoid force pushing to your repo)

lindsayad replied:
https://github.com/idaholab/moose/pull/16117/commits/85350e9d2c7f3d54b5d03424d280968f0648ca90

YaqiWang replied:
Force push should be fine. I stopped working on this branch. It is yours.

YaqiWang commented on framework/include/base/NeighborResidualObject.h:
```diff
@@ -0,0 +1,34 @@
+//* This file is part of the MOOSE framework
+//* https://www.mooseframework.org
+//*
```
This should be merged into previous commit. The commit message is not right.

lindsayad replied:
You've never used `git commit --fixup` and `git rebase --autosquash` ? It is easy to autosquash all these commits later. If you give me permission to force push willy nilly to this branch, then I won't push up `fixup` commits anymore

YaqiWang replied:
Oh, you cannot do force push?

YaqiWang replied:
My repository is public. I thought anyone can force push to my branches. If not, where should set this up?

YaqiWang replied:
Oh maybe feel free to create a new PR with your own branch.

lindsayad replied:
I can. Was just not doing it at first out of courtesy to you. But now that you've said this branch is all mine, I'll force push :+1:

YaqiWang commented:
Really nice. I think we should add some inline comment before `ResidualObject::computeOffdiagJacobian` to give reasons why we design the interface in this way.

YaqiWang commented on framework/include/base/ResidualObject.h:
```diff
@@ -67,7 +66,7 @@ class ResidualObject : public MooseObject,
    * Computes Jacobian block of the variable of this object with respect to a field variable
    * @param jvar The field variable
```
It will be nice to add some docs here to explain why we should to pass variable number here. Note this line of doxygen comment is not accurate anymore.

lindsayad committed:
Add some backwards compatibility

- Retain `ODEKernel::computeQpOffDiagJacobian`

lindsayad commented:
Patches:

- [x] Griffin
- [ ] THM
- [ ] Mastodon

YaqiWang commented:
@fdkong I guess this can be merged.

fdkong commented:
> @fdkong I guess this can be merged.

It is still failing. We need to force merge failing things for integration?

loganharbour commented:
Please don't merge until we at least have fixes ready to be merged up for the affected apps.

lindsayad commented:
See https://github.com/idaholab/moose/pull/16117#issuecomment-725125095. This will probably be one to two days work to get this all integrated, but I think it's well worth it. It's impossible to be backwards compatible when changing APIs that currently take a non-const reference to taking a const reference (at least I haven't thought of a way to do it)

loganharbour commented:
> See [#16117 (comment)](https://github.com/idaholab/moose/pull/16117#issuecomment-725125095). 

I saw - I made the comment because I believe you're missing one or two

lindsayad commented:
> I saw - I made the comment because I believe you're missing one or two

Which ones? They're failures independent of THM/Griffin?

loganharbour commented:
> They're failures independent of THM/Griffin?

Disregard, I failed to note the dependency

YaqiWang commented:
@lindsayad Looks like that we still need to check the following three:
- [ ] Pronghorn
- [ ] Relap-7
- [ ] Sockeye

I can help Pronghorn and Relap-7.

loganharbour commented:
> @lindsayad Looks like that we still need to check the following three:

Those are good. Look at the conversation above about THM/Griffin dependency.

YaqiWang commented:
Where is the conversation?

loganharbour reviewed (approved):
For @lindsayad's commits

lindsayad reviewed (approved):
Removing my un-approval. @fdkong you want to look once more? You are the only one with outstanding requested changes

lindsayad reviewed (approved):
Removing my un-approval. @fdkong you want to look once more? You are the only one with outstanding requested changes

fdkong commented:
> Removing my un-approval. @fdkong you want to look once more? You are the only one with outstanding requested changes

I am looking at it now

fdkong commented:
OK, merged. @lindsayad Could you follow up the integration process?

YaqiWang commented:
Congratulation! ;-)

lindsayad commented:
> OK, merged. @lindsayad Could you follow up the integration process?

Yes
